File: exercism/elixir/leap/leap.exs
defmodule Year do
  @doc """
  Returns whether 'year' is a leap year.

  A leap year occurs:

  on every year that is evenly divisible by 4
  except every year that is evenly divisible by 100
  except every year that is evenly divisible by 400.
  """
  @spec leap_year?(non_neg_integer) :: boolean
  def leap_year?(year) do
    is = fn(y) -> rem(year, y) === 0 end
    is_not = fn(a) -> not is.(a) end

    cond do
      is.(4) and is.(100) and is_not.(400)-> false
      is.(4) -> true
      is.(400) -> true
      true -> false
    end
  end
end

File: exercism/scala/leap/src/main/scala/Leap.scala
object Year {
  def apply(year: Int) = new Year(year)
}

class Year(year: Int) {
  def evenly(n: Int) = year % n == 0
  def isLeap: Boolean = (evenly(400), evenly(100), evenly(4)) match {
    case (true, _, _) => true
    case (_, false, true) => true
    case _ => false
  }
}
